### PR TITLE
[Win] Refactor event system

### PR DESCRIPTION
**Pull request by LtSnail**
Using std::variant and std::visitor instead of EventDispatcher.
Improved EventSystem design

### CODE REVIEW

File: Engine/Fuego/Events/ApplicationEvent.h [hunk 1]
```diff
@@ -13,11 +13,11 @@ class FUEGO_API WindowResizeEvent : public Event
     {
     }
 
-    inline unsigned int GetWidth() const
+    unsigned int GetWidth() const
```
Comment (artem-zyktin) [inline Engine/Fuego/Events/ApplicationEvent.h:16]: In general I would recommend you to use fix sized type aliases like `uint32_t`

File: Engine/Fuego/Events/ApplicationEvent.h [hunk 2]
```diff
@@ -13,11 +13,11 @@ class FUEGO_API WindowResizeEvent : public Event
...
     {
         return m_Width;
     }
-    inline unsigned int GetHeight() const
+    unsigned int GetHeight() const
     {
         return m_Height;
```
Comment (artem-zyktin) [inline Engine/Fuego/Events/ApplicationEvent.h:22]: Do not make implementation of non-template classes inside a header. Any implementation changes inside a header lead to recompilation of sources which depend on this header.

File: Engine/Fuego/Events/Event.h [hunk 3]
```diff
@@ -1,90 +1,35 @@
...
-    const char* GetName() const override    \
-    {                                       \
-        return #type;                       \
-    }
-
-#define EVENT_CLASS_CATEGORY(category)    \
-    int GetCategoryFlags() const override \
-    {                                     \
-        return category;                  \
+#define EVENT_CLASS_TYPE(type)           \
```
Comment (artem-zyktin) [inline Engine/Fuego/Events/Event.h:8]: Do you really wanna this macro?

File: Engine/Fuego/Windows/EventQueueWin.cpp [hunk 4]
```diff
@@ -6,24 +6,24 @@ void EventQueueWin::Update()
 {
 }
 
-std::shared_ptr<Event> EventQueueWin::Front()
+std::unique_ptr<EventVariant>& EventQueueWin::Front()
```
Comment (artem-zyktin) [inline Engine/Fuego/Windows/EventQueueWin.cpp:9]: I can't recomend you to use unique_ptr referencves, epecially non const: 
1. It breaks down semantics of uniqueness of resource ownership.
2. What happen if recieving code call `std::move` or `reset`? If you afraid of cyclic references - use `weak_ptr` and `shared_ptr`.

**Review by artem-zyktin — CHANGES_REQUESTED** [review]

File: Engine/Fuego/Events/ApplicationEvent.h (hunk 1 above)
Comment (LtSnail) [reply Engine/Fuego/Events/ApplicationEvent.h:16]: Fixed

File: Engine/Fuego/Events/ApplicationEvent.h (hunk 2 above)
Comment (LtSnail) [reply Engine/Fuego/Events/ApplicationEvent.h:22]: Discussed in our chat

File: Engine/Fuego/Windows/EventQueueWin.cpp (hunk 4 above)
Comment (LtSnail) [reply Engine/Fuego/Windows/EventQueueWin.cpp:9]: Changed back to UPtr

File: Engine/Fuego/Events/Event.h (hunk 3 above)
Comment (LtSnail) [reply Engine/Fuego/Events/Event.h:8]: Fixed

**Review by artem-zyktin — APPROVED** [review]

**Pull request merged**